File: built-in plugins dsp/mb-distortion.cpp
/**
 * MB Distortion
 * Category : effect
 * Type     : distortion
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Multi-mode distortion with tube, tape, and digital saturation
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_DISTORTION_H
#define MB_DISTORTION_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbDistortion : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-distortion";
    static constexpr const char* PLUGIN_NAME    = "MB Distortion";
    static constexpr const char* PLUGIN_TYPE    = "distortion";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float drive = 0.5f;  // range [0, 1]
    float tone = 0.5f;  // range [0, 1]
    float output = 0f;  // range [-24, 12]
    float mix = 1.0f;  // range [0, 1]
    float bias = 0f;  // range [-1, 1]
    };

    MbDistortion() = default;
    ~MbDistortion() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.drive = std::clamp(params.drive, 0f, 1f);
        params.tone = std::clamp(params.tone, 0f, 1f);
        params.output = std::clamp(params.output, -24f, 12f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        params.bias = std::clamp(params.bias, -1f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Distortion
        return input;
    }
};

#endif // MB_DISTORTION_H

File: built-in plugins dsp/mb-distortion.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbDistortionPlugin: PluginDefinition = {
  id: "mb-distortion",
  slug: "mb-distortion",
  name: "MB Distortion",
  category: "effect",
  type: "distortion",
  version: "1.0.0",
  description: "Multi-mode distortion with tube, tape, and digital saturation",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "drive",
      name: "Drive",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "tone",
      name: "Tone",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "output",
      name: "Output Level",
      type: "float",
      defaultValue: 0,
      minValue: -24,
      maxValue: 12,
      automatable: true,
    },
    {
      id: "mix",
      name: "Mix",
      type: "float",
      defaultValue: 1.0,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "bias",
      name: "Bias",
      type: "float",
      defaultValue: 0,
      minValue: -1,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { drive: 0.5, tone: 0.5, output: 0, mix: 1.0, bias: 0 },
};

export default MbDistortionPlugin;
